storage/src/test/scala/io/delta/storage/integration/S3LogStoreUtilIntegrationTest.scala

```diff
@@ -0,0 +1,107 @@
+
+import io.delta.storage.internal.S3LogStoreUtil
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.fs.s3a.{S3AFileSystem, UploadInfo}
+import org.scalatest.Tag
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalactic.source
+
+import java.net.URI
+import java.nio.file.Paths
+import scala.math.max
+import scala.math.ceil
+import scala.math.round
+
+class S3LogStoreUtilIntegrationTest extends AnyFunSuite {
+  private val runIntegrationTests: Boolean =
+    Option(System.getenv("S3_LOG_STORE_UTIL_TEST_ENABLED")).exists(_.toBoolean)
+  private val bucket = System.getenv("S3_LOG_STORE_UTIL_TEST_BUCKET")
+  private val keyPrefix = System.getenv("S3_LOG_STORE_UTIL_TEST_KEY_PREFIX")
+  private lazy val fs: S3AFileSystem = {
+    val fs = new S3AFileSystem()
+    fs.initialize(new URI(s"s3a://$bucket"), configuration)
+    fs
+  }
+  private val configuration = new Configuration()
+  configuration.set( // for local testing only
+    "fs.s3a.aws.credentials.provider",
+    "com.amazonaws.auth.profile.ProfileCredentialsProvider"
+  )
+
+  private val file = Paths.get(getClass.getClassLoader.getResource("sample.json").getPath).toFile
+
+  private def touch(key: String): UploadInfo =
+    fs.putObject(fs.newPutObjectRequest(key, fs.newObjectMetadata(), file))
+
+  private def key(table: String, version: Int): String =
+    s"$keyPrefix/$table/_delta_log/%020d.json".format(version)
+
+  private def path(table: String, version: Int): Path =
+    new Path(s"s3a://$bucket/${key(table, version)}")
+
+  private def version(path: Path): Int
+  = path.getName.takeWhile(_ != '.').toInt
+
+  private val integrationTestTag = Tag("IntegrationTest")
+
+  def integrationTest(name: String)(testFun: => Any)(implicit pos: source.Position): Unit =
+    if (runIntegrationTests) test(name, integrationTestTag)(testFun)(pos)
+
+  integrationTest("setup delta logs") {
+    val uploads = Seq(
+      touch(s"$keyPrefix/empty/some.json"),
+      touch(s"$keyPrefix/small/_delta_log/%020d.json".format(1)),
+      touch(s"$keyPrefix/small/_before/some.json"),
+      touch(s"$keyPrefix/small/_right_after/some.json")) ++
+      (1 to 10).map(v => touch(s"$keyPrefix/medium/_delta_log/%020d.json".format(v))) ++
+      (1 to 1000).map(v => touch(s"$keyPrefix/large/_delta_log/%020d.json".format(v))) ++
+      (1 to 10000).map(v => touch(s"$keyPrefix/xlarge/_delta_log/%020d.json".format(v)))
+    uploads.foreach(_.getUpload.waitForUploadResult())
+  }
+
+  integrationTest("empty") {
+    val resolvedPath = path("empty", 0)
+    val response = S3LogStoreUtil.s3ListFrom(fs, resolvedPath, resolvedPath.getParent)
+    assert(response.isEmpty)
+  }
+
+  integrationTest("small") {
+    Seq(0, 1, 2, 3).foreach(v => {
+      val resolvedPath = path("small", v)
+      val response = S3LogStoreUtil.s3ListFrom(fs, resolvedPath, resolvedPath.getParent)
+      assert((max(1, v) to 1) == response.map(r => version(r.getPath)).toSeq)
```

Comment:
add a comment describing the semantics of this. help future readers easily tell what we are checking for here.

Reply:
Added here https://github.com/delta-io/delta/pull/1210/commits/ea6ca2459e638fc1e2f5d8547e9ba2d23c00d199